pyquil/operator_estimation.py

```diff
@@ -0,0 +1,428 @@
+
+    def __repr__(self):
+        return f'Experiment[{self}]'
+
+    def serializable(self):
+        return str(self)
+
+    @classmethod
+    def from_str(cls, s: str):
+        """The opposite of str(exp)"""
+        instr, outstr = s.split('\u2192')
+        return cls(in_operator=PauliTerm.from_str(instr),
+                   out_operator=PauliTerm.from_str(outstr))
+
+
+def _abbrev_program(program: Program, abbrev_after=10):
+    program_lines = program.out().splitlines()
+    if abbrev_after is not None and len(program_lines) > abbrev_after:
+        first_n = abbrev_after // 2
+        last_n = abbrev_after - first_n
+        excluded = len(program_lines) - abbrev_after
+        program_lines = (program_lines[:first_n] + [f'... {excluded} hidden ...']
+                         + program_lines[-last_n:])
+
+    return '; '.join(program_lines)
+
+
+class ExperimentSuite:
+    """
+    A whole set of tomography-like experiments
+
+    Many near-term quantum algorithms involve measuring an ansatz state with respect to
+    many different observables. This class contains all (in_pauli, out_pauli) pairs for
+    a given ansatz preparation program.
+
+    Experiments belonging to a shared tensor product basis (TPB) can (optionally) be estimated
+    simultaneously. Therefore, this class is backed by a list of list of Experiments.
+    Experiments sharing an inner list will be estimated simultaneously. If you don't want this,
+    provide a list of length-1-lists. As a convenience, if you pass a 1D list to the constructor
+    it will expand it to a list of length-1-lists.
+
+    Please see :py:func:`group_experiments` for a function that will automatically process an
+    ExperimentSuite to group Experiments sharing a TPB.
+    """
+
+    def __init__(self,
+                 experiments: Union[List[Experiment], List[List[Experiment]]],
+                 program: Program,
+                 qubits: List[int]):
+        if len(experiments) == 0:
+            experiments = []
+        else:
+            if isinstance(experiments[0], Experiment):
+                # convenience wrapping in lists of length 1
+                experiments = [[exp] for exp in experiments]
+
+        self._experiments = experiments  # type: List[List[Experiment]]
+        self.program = program
+        self.qubits = qubits
+
+    def __len__(self):
+        return len(self._experiments)
+
+    def __getitem__(self, item):
+        return self._experiments[item]
+
+    def __setitem__(self, key, value):
+        self._experiments[key] = value
+
+    def __delitem__(self, key):
+        self._experiments.__delitem__(key)
+
+    def __iter__(self):
+        yield from self._experiments
+
+    def __reversed__(self):
+        yield from reversed(self._experiments)
+
+    def __contains__(self, item):
+        return item in self._experiments
+
+    def append(self, exp):
+        return self._experiments.append(exp)
+
+    def count(self, exp):
```

Comment:
This relates to my comment above about `__contains__`. An ExperimentSuite cannot count Experiments using this, only lists of Experiments. For example, if we did
```
import pyquil.operator_estimation as oe
import pyquil.paulis as pl
expt1 = oe.Experiment(pl.sX(1) * pl.sZ(0), pl.sX(2) * pl.sY(1) * pl.sZ(0))
expt2 = oe.Experiment(pl.sZ(1) * pl.sX(0), pl.sY(2) * pl.sZ(1) * pl.sX(0))
expt3 = oe.Experiment(pl.sX(2) * pl.sY(1) * pl.sZ(0), pl.sZ(2) * pl.sX(1) * pl.sY(0))
expt_list = [expt1, expt2, expt3]
expt_suite = oe.ExperimentSuite(expt_list, Program(H(0), H(1), H(2)), [0, 1, 2])
```
and then do
```
expt_suite.count(expt1)
```
we'd get `0`. On the other hand, if we did
```
expt_suite.count([expt1])
```
we'd get `1`.

I suppose this is OK as long as we make the user well aware of this.

Reply:
I'm still wondering if it's worth adding something like `if isinstance(input_expt, Experiment): return flattened(self._experiments.count(input_expt)`to the `count(..)` function. It would still be valuable to just count for lists (of Experiments) though, since we might want to know whether a particular group of Experiments exists in some ExperimentSuite after term grouping.